Added an interactive Bokeh widget to create an interactive echelle diagram.

christinahedges opened the pull request:
I added a new widget that lets users create an interactive echelle diagram. This is very useful for exploring the data and hunting for oscillation modes. Because it's bokeh, it's very responsive. This is totally based on the idea and success of @danhey's tool for interactive echelle diagrams, but uses lightkurve and bokeh.

![image](https://user-images.githubusercontent.com/14965634/68521192-6c419580-0253-11ea-9659-6bf9b7a057ac.png)

`interact_echelle` takes the same keyword arguments as `plot_echelle`. 

### Changes:

* Refactored the `plot_echelle` function into a plotting function and a helper function
* Added some bokeh helper functions to create the widget
* Added an `interact_echelle` function.


### API:

```python
seis = lc.to_periodogram(freq_unit='microhertz').to_seismology()
seis.interact_echelle()
```

christinahedges commented:
I do need to make the slider a little more responsive, currently it's not as fine as I'd like.

danhey commented:
Hi Christina, I have nothing to add except that this is very cool and I'm glad to see it become a feature in Lightkurve!

ojhall94 commented:
Hi Christina,

Thanks for doing this! I'll try and give it a look over tomorrow and provide some useful comments!

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+        if smooth_filter_width:
+            pgsmooth = self.periodogram.smooth(filter_width=smooth_filter_width)
+            freq = pgsmooth.frequency  # Makes code below more readable below
+            power = pgsmooth.power     # Makes code below more readable below
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'power'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'power'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+            ll_button.on_click(go_left_by_one)
+
+            widgets_and_figures = layout([fig_tpf, [Spacer(height=20), stretch_slider]],
+                                         [ll_button, Spacer(width=30), l_button, Spacer(width=25), dnu_slider, Spacer(width=30), r_button, Spacer(width=23), rr_button])
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [line too long (168 > 159 characters)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [line too long (168 > 159 characters)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -12,9 +12,22 @@
+    import bokeh  # Import bokeh first so we get an ImportError we can catch
+    from bokeh.io import show, output_notebook, push_notebook
+    from bokeh.plotting import figure
+    from bokeh.models import LogColorMapper, Slider, RangeSlider, \
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -12,9 +12,22 @@
+    from bokeh.models import LogColorMapper, Slider, RangeSlider, \
+                             Span, LogTicker, Range1d, Button
+    from bokeh.layouts import layout, Spacer
+except:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [No exception type(s) specified](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [No exception type(s) specified](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [No exception type(s) specified](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [No exception type(s) specified](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -12,9 +12,22 @@
+# Import the optional Bokeh dependency, or print a friendly error otherwise.
+try:
+    import bokeh  # Import bokeh first so we get an ImportError we can catch
+    from bokeh.io import show, output_notebook, push_notebook
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.io.push_notebook' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused push_notebook imported from bokeh.io](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused push_notebook imported from bokeh.io](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.io.push_notebook' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.io.push_notebook' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused push_notebook imported from bokeh.io](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.io.push_notebook' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused push_notebook imported from bokeh.io](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -97,19 +110,11 @@ def _validate_deltanu(self, deltanu):
-        value or a low signal to noise case.
-
-        This method is adapted from work by Daniel Hey & Guy Davies.
+    def _clean_echelle(self, deltanu=None, numax=None,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+        return fig, stretch_slider
+
+
+    def interact_echelle(self, notebook_url='localhost:8888', **kwargs):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology.interact_echelle is too complex (16)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -12,9 +12,22 @@
 
+# Import the optional Bokeh dependency, or print a friendly error otherwise.
+try:
+    import bokeh  # Import bokeh first so we get an ImportError we can catch
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+            power = pgsmooth.power     # Makes code below more readable below
+        else:
+            freq = self.periodogram.frequency  # Makes code below more readable
+            power = self.periodogram.power     # Makes code below more readable
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'power' is assigned to but never used](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'power' is assigned to but never used](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+            power = self.periodogram.power     # Makes code below more readable
+
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+        """
+
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+                """Callback to take action when dnu slider changes"""
+                dnu = SeismologyQuantity(quantity=dnu_slider.value*u.microhertz,
+                                                       name='deltanu', method='echelle')
+                ep, x_f, y_f = self._clean_echelle(deltanu=dnu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'y_f'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'y_f'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+        return fig, stretch_slider
+
+
+    def interact_echelle(self, notebook_url='localhost:8888', **kwargs):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology.interact_echelle is too complex (16)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

christinahedges replied:
bite me.

ojhall94 commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
+                                                    maximum_frequency=maximum_frequency,
+                                                    **kwargs)
+                fig_tpf.select('img')[0].data_source.data['image'] = [ep.value]
+                fig_tpf.xaxis.axis_label = r'Frequency / {:.3f} Mod. 1'.format(dnu)
```
We should use the same unit axis labels as we use in `periodogram`, and the same phrasing as in `plot_echelle()` e.g.: 

```
axis_label = rf" Frequency mod. {dnu.value:.3f} [{deltanu.unit.to_string('latex')}]"
```

christinahedges replied:
Hi Ollie, the x axis is actually different in the interact echelle diagram. In order for the slider to be "smooth" the x axis is 

`frequency/dnu %  1`

Which means the x axis spans 0 to 1 in phase

In the original echelle diagram the x axis is

`frequency %  dnu`

So it goes from 0 to dnu.

ojhall94 commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +353,170 @@ def plot_echelle(self, deltanu=None, numax=None,
 
             cbar.set_label(ylabel)
             ax.set_xlabel(r'Frequency mod. {:.2f}'.format(deltanu))
```
We should make sure the unit is done properly in LaTeX + is consistent with our other plots, e.g.

```
axis_label = rf" Frequency mod. {dnu.value:.3f} [{deltanu.unit.to_string('latex')}]"
```

christinahedges replied:
Bokeh can't render latex, so I've omitted this.

barentsen commented:
This looks really awesome and promising!

I tried it using the code below, which seemed a bit fickle (the diagram goes blank at the start and after playing with the sliders):

```python
import lightkurve as lk
search = lk.search_lightcurvefile('KIC10963065', cadence='short', mission='Kepler')
files = search[1:10].download_all()
lc = files.PDCSAP_FLUX.stitch().remove_nans()
seis = lc.to_periodogram(freq_unit='microhertz').to_seismology()
seis.interact_echelle('http://localhost:8889')
```

christinahedges commented:
Hey Ollie,

Thanks for a great review!! Here are some responses:

> Is it possible to build unit tests for the bokeh widget?

@gully and I think this is not really possible, but we're open to ideas on this!

> Does the vertical slider change the vmin and vmax of the signal-to-noise? It would be more useful (and more intuitive) I think if instead it selected the minimum and maximum frequency.

I had thought it was easier to set up the minimum/maximum frequency in the call to the function, but i can make a separate slider for that. The slider for the vmax/vmin is important to get to those low signal to noise peaks.

> Can we set the cmap to whatever the default is for plot_echelle(), and maintain its functionality as a kwarg in interact_echelle()?

> If I pass a cmap kwarg, it still shows the plot, but shouts errors at me if I slide DeltaNu around

> It would be really neat for consistency if we could pass a cmap into the Bokeh widget

I can set the same cmap, but I don't think I can make it a kwarg because bokeh strings for cmaps are different to matplotlib strings to cmaps. This is why it's breaking for yiou.

> Similarly it would be really tidy if we could fix the size of the widget to the same figure size we use for the basic plot_echelle()

I can make it a more consistent size, no problem!

Thanks for the review I'll get to work on your other comments.

christinahedges commented:
> I tried it using the code below, which seemed a bit fickle (the diagram goes blank at the start and after playing with the sliders):

This appears to be a linux issue.

christinahedges commented on lightkurve/seismology/core.py:
```diff
@@ -12,9 +12,22 @@
+                             Span, LogTicker, Range1d, Button
+    from bokeh.layouts import layout, Spacer
+except:
+    # Nice error when interact_echelle is called.
```
Hi Ollie this error message is shown in line 431.

christinahedges commented:
I've addressed most of @ojhall94's comments, and those I couldn't I've responded to on this PR. The new interact echelle now looks like this:

![image](https://user-images.githubusercontent.com/14965634/69107672-01e4de00-0a27-11ea-95a7-8277e59614f0.png)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+        """
+
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+            def go_right_by_one():
+                """Step forward in time by a single cadence"""
+                existing_value = dnu_slider.value
+                if existing_value < maxdnu:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [undefined name 'maxdnu'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+            ll_button.on_click(go_left_by_one)
+
+            widgets_and_figures = layout([fig_tpf, [Spacer(height=20), stretch_slider]],
+                                         [ll_button, Spacer(width=30), l_button, Spacer(width=25), dnu_slider, Spacer(width=30), r_button, Spacer(width=23), rr_button])
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [line too long (168 > 159 characters)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+        if smooth_filter_width:
+            pgsmooth = self.periodogram.smooth(filter_width=smooth_filter_width)
+            freq = pgsmooth.frequency  # Makes code below more readable below
+            power = pgsmooth.power     # Makes code below more readable below
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'power'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+        return fig, stretch_slider
+
+
+    def interact_echelle(self, notebook_url='localhost:8888', **kwargs):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology.interact_echelle is too complex (16)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+            power = pgsmooth.power     # Makes code below more readable below
+        else:
+            freq = self.periodogram.frequency  # Makes code below more readable
+            power = self.periodogram.power     # Makes code below more readable
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'power' is assigned to but never used](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -12,9 +12,22 @@
+    import bokeh  # Import bokeh first so we get an ImportError we can catch
+    from bokeh.io import show, output_notebook
+    from bokeh.plotting import figure
+    from bokeh.models import LogColorMapper, Slider, RangeSlider, \
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Range1d imported from bokeh.models](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh.models.Span' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+                """Callback to take action when dnu slider changes"""
+                dnu = SeismologyQuantity(quantity=dnu_slider.value*u.microhertz,
+                                                       name='deltanu', method='echelle')
+                ep, x_f, y_f = self._clean_echelle(deltanu=dnu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'y_f'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,211 @@ def plot_echelle(self, deltanu=None, numax=None,
+            power = self.periodogram.power     # Makes code below more readable
+
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+        return fig, stretch_slider
+
+
+    def interact_echelle(self, notebook_url='localhost:8888', **kwargs):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology.interact_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+                """Callback to take action when dnu slider changes"""
+                dnu = SeismologyQuantity(quantity=dnu_slider.value*u.microhertz,
+                                                       name='deltanu', method='echelle')
+                ep, x_f, y_f = self._clean_echelle(deltanu=dnu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'y_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+        """
+
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -276,11 +354,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+            freq = self.periodogram.frequency  # Makes code below more readable
+
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

christinahedges commented:
> I request 3 changes:

> Please fix the issues identified by codacy above ☝️ .
> Please git rebase master because we fixed a few bugs in the past 48h.
> The current version of this branch seems to crash for me:

1. I fixed most of the codacy changes, some of them are crying when we reimport bokeh so that users without bokeh will have a useful error message

2. I rebased

3. I resolved your crash, it was missing a unit. Thanks for the check!

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +358,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+                """Callback to take action when dnu slider changes"""
+                dnu = SeismologyQuantity(quantity=dnu_slider.value*u.microhertz,
+                                                       name='deltanu', method='echelle')
+                ep, x_f, y_f = self._clean_echelle(deltanu=dnu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'y_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +358,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+        return fig, stretch_slider
+
+
+    def interact_echelle(self, notebook_url='localhost:8888', **kwargs):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology.interact_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +358,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+        """
+
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +358,212 @@ def plot_echelle(self, deltanu=None, numax=None,
+            freq = self.periodogram.frequency  # Makes code below more readable
+
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -98,18 +111,11 @@ def _validate_deltanu(self, deltanu):
-        value or a low signal to noise case.
-
-        This method is adapted from work by Daniel Hey & Guy Davies.
+    def _clean_echelle(self, deltanu=None, numax=None,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +357,167 @@ def plot_echelle(self, deltanu=None, numax=None,
+            power = pgsmooth.power     # Makes code below more readable below
+        else:
+            freq = self.periodogram.frequency  # Makes code below more readable
+            power = self.periodogram.power     # Makes code below more readable
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'power' is assigned to but never used](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +357,167 @@ def plot_echelle(self, deltanu=None, numax=None,
+        """
+
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +357,167 @@ def plot_echelle(self, deltanu=None, numax=None,
+        if smooth_filter_width:
+            pgsmooth = self.periodogram.smooth(filter_width=smooth_filter_width)
+            freq = pgsmooth.frequency  # Makes code below more readable below
+            power = pgsmooth.power     # Makes code below more readable below
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'power'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +357,167 @@ def plot_echelle(self, deltanu=None, numax=None,
+            ll_button.on_click(go_left_by_one)
+
+            widgets_and_figures = layout([fig_tpf, [Spacer(height=20), stretch_slider]],
+                                         [ll_button, Spacer(width=30), l_button, Spacer(width=25), dnu_slider, Spacer(width=30), r_button, Spacer(width=23), rr_button])
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [line too long (168 > 159 characters)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +357,167 @@ def plot_echelle(self, deltanu=None, numax=None,
+            power = self.periodogram.power     # Makes code below more readable
+
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +354,175 @@ def plot_echelle(self, deltanu=None, numax=None,
+            of the form "localhost:8888", then "http" will be used.
+        """
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +354,175 @@ def plot_echelle(self, deltanu=None, numax=None,
+        else:
+            freq = self.periodogram.frequency  # Makes code below more readable
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -12,8 +13,19 @@
+    from bokeh.plotting import figure
+    from bokeh.models import LogColorMapper, Slider, RangeSlider, Button
+    from bokeh.layouts import layout, Spacer
+except:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [No exception type(s) specified](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
````diff
@@ -12,8 +13,19 @@
+# Import the optional Bokeh dependency required by ``interact_echelle```,
+# or print a friendly error otherwise.
+try:
+    import bokeh  # Import bokeh first so we get an ImportError we can catch
````
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['bokeh' imported but unused](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import bokeh](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -98,18 +110,11 @@ def _validate_deltanu(self, deltanu):
-        value or a low signal to noise case.
-
-        This method is adapted from work by Daniel Hey & Guy Davies.
+    def _clean_echelle(self, deltanu=None, numax=None,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Seismology._clean_echelle is too complex (16)](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +349,175 @@ def plot_echelle(self, deltanu=None, numax=None,
+        else:
+            freq = self.periodogram.frequency  # Makes code below more readable
+
+        ep, x_f, y_f = self._clean_echelle(deltanu=deltanu,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'x_f'](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)

barentsen commented on lightkurve/seismology/core.py:
```diff
@@ -280,11 +349,175 @@ def plot_echelle(self, deltanu=None, numax=None,
+            of the form "localhost:8888", then "http" will be used.
+        """
+        try:
+            import bokeh
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [redefinition of unused 'bokeh' from line 21](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4468966)
